### PR TITLE
Fix CocosApplication doesn't unregister engine event when restarting.

### PR DESCRIPTION
**Pull request by dumganhar**
-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/core/event/CallbacksInvoker.h [hunk 1]
```diff
@@ -42,8 +42,24 @@ namespace cc {
...
+        }
+        value = rhs.value;
+        return *this;
+    }
+
+    uint32_t value{0};
+};
+
 struct CallbackInfoBase {
-    using ID = uint32_t;
```
Comment (dumganhar) [inline native/cocos/core/event/CallbacksInvoker.h:46]: Put ID as a strong type `struct CallbackID` to distinguish with the key which is also a uint32_t typedef.

File: native/tools/tojs/scene.ini [hunk 2]
```diff
@@ -77,7 +77,7 @@ skip = Pass::[getBlocks initPassFromTarget],
...
+       Node::[setLayerPtr setUIPropsTransformDirtyCallback rotate$ setUserData getUserData getChildren rotateForJS setScale$ setRotation$ setRotationFromEuler$ setPosition$ isActiveInHierarchy setActiveInHierarchy setActiveInHierarchyPtr setRTS$ findComponent findChildComponent findChildComponents addComponent removeComponent getComponent getComponents getComponentInChildren getComponentsInChildren checkMultipleComp getEventProcessor dispatchEvent hasEventListener getUIProps getPosition getRotation getScale getEulerAngles getForward getUp getRight getWorldPosition getWorldRotation getWorldScale getWorldMatrix getWorldRS getWorldRT inverseTransformPoint ^on$ ^off$],
```
Comment (dumganhar) [inline native/tools/tojs/scene.ini:80]: Node::on/off should not be bound since they are done in ts.

**Review by qiuguohua — APPROVED** [review]

File: native/cocos/application/CocosApplication.cpp [hunk 3]
```diff
@@ -45,15 +45,24 @@ CocosApplication::CocosApplication() {
...
-    _engine->off(BaseEngine::ON_CLOSE);
+    unregisterAllEngineEvents();
+}
+
+void CocosApplication::unregisterAllEngineEvents() {
+    if (_engine != nullptr) {
+        _engine->offAll(BaseEngine::ON_START);
+        _engine->offAll(BaseEngine::ON_RESUME);
+        _engine->offAll(BaseEngine::ON_PAUSE);
+        _engine->offAll(BaseEngine::ON_CLOSE);
```
Comment (dumganhar) [inline native/cocos/application/CocosApplication.cpp:56]: Use offAll to unregister all callbacks with event key.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

File: native/cocos/core/scene-graph/NodeEventProcessor.cpp [hunk 4]
```diff
@@ -380,6 +380,29 @@ void NodeEventProcessor::off(const CallbacksInvoker::KeyType &type, CallbackInfo
...
+    //                event::EventManager::getInstance()->removeEventListener(_touchListener);
+    //                _touchListener = nullptr;
+    //            }
+    //        } else if (mouseEventExist) {
+    //            if (_mouseListener && !checkListeners(_node, MOUSE_EVENTS)) {
+    //                event::EventManager::getInstance()->removeEventListener(_mouseListener);
+    //                _mouseListener = nullptr;
+    //            }
+    //        }
+    //    } else
```
Comment (minggo) [inline native/cocos/core/scene-graph/NodeEventProcessor.cpp:400]: debug code?

**Review by novain1102 — APPROVED** [review]

File: native/cocos/core/event/CallbacksInvoker.h [hunk 5]
```diff
@@ -42,8 +42,24 @@ namespace cc {
 #define CC_CALLBACK_INVOKE_2(__selector__, __target__, Arg0, Arg1, ...)       std::function<void(Arg0, Arg1)>(std::bind(&__selector__, __target__, std::placeholders::_1, std::placeholders::_2, ##__VA_ARGS__)), __target__
 #define CC_CALLBACK_INVOKE_3(__selector__, __target__, Arg0, Arg1, Arg2, ...) std::function<void(Arg0, Arg1, Arg2)>(std::bind(&__selector__, __target__, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, ##__VA_ARGS__)), __target__
 
+struct CallbackID {
+    CallbackID() = default;
+    explicit CallbackID(uint32_t v) : value(v) {}
+    CallbackID(const CallbackID &rhs) { value = rhs.value; }
```
Comment (minggo) [inline native/cocos/core/event/CallbacksInvoker.h]: Can also use initialize list.

File: native/cocos/core/scene-graph/NodeEventProcessor.cpp (hunk 4 above)
Comment (dumganhar) [reply native/cocos/core/scene-graph/NodeEventProcessor.cpp:400]: It isn't debug code.
At the beginning of developing v3.6, we make the whole engine running on c++ without ts logic. 
NodeEventProcessor internally depends on system event (touch, mouse) handling which is not needed now, so we commented them temporarily, maybe we need to uncomment them in the further as more and more logic moved to cpp side.

File: native/cocos/core/event/CallbacksInvoker.h (hunk 5 above)
Comment (dumganhar) [reply native/cocos/core/event/CallbacksInvoker.h]: The code like `CallbackID cbID{100}` works for now. Seems doesn't need to explicit initialize_list constructor.

Comment (minggo) [reply native/cocos/core/event/CallbacksInvoker.h]: What i mean is
```c++
CallbackID(const CallbackID &rhs): value(rhs.value) { }
```

Comment (dumganhar) [reply native/cocos/core/event/CallbacksInvoker.h]: Okay.

Comment (dumganhar) [reply native/cocos/core/event/CallbacksInvoker.h]: Done

File: native/cocos/core/event/CallbacksInvoker.h [hunk 6]
```diff
@@ -42,8 +42,24 @@ namespace cc {
...
+    CallbackID &operator=(const CallbackID &rhs) {
+        if (this == &rhs) {
+            return *this;
+        }
+        value = rhs.value;
+        return *this;
+    }
+
+    uint32_t value{0};
+};
```
Comment (PatriceJiang) [inline native/cocos/core/event/CallbacksInvoker.h]:
```c++
struct CallbackID {
    uint32_t value{0};
};
```
All functions should use default impl?

**Review by minggo — APPROVED** [review]

File: native/cocos/core/event/CallbacksInvoker.h (hunk 6 above)
Comment (dumganhar) [reply native/cocos/core/event/CallbacksInvoker.h]: Done

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Review by PatriceJiang — APPROVED** [review]

File: native/cocos/application/CocosApplication.cpp [hunk 7]
```diff
@@ -45,15 +45,24 @@ CocosApplication::CocosApplication() {
 }
 
 CocosApplication::~CocosApplication() {
-    _engine->off(BaseEngine::ON_RESUME);
```
Comment (dumganhar) [inline native/cocos/application/CocosApplication.cpp:48]: Miss off ON_START

**Pull request merged**